pkg/controller/service/service_controller.go

```diff
@@ -719,47 +608,114 @@ func (s *ServiceController) syncService(key string) error {
+		return nil
+	}
+
+	if err != nil {
 		glog.Infof("Unable to retrieve service %v from store: %v", key, err)
-	default:
-		cachedService = s.cache.getOrCreate(key)
-		err = s.processServiceUpdate(cachedService, service, key)
+		return err
 	}
 
-	return err
+	if isDeletionCandidate(service) {
+		err := s.processServiceDeletion(service, key)
+		if err != nil {
+			return err
+		}
+
+		service, err = s.removeFinalizer(service)
+		if err != nil {
+			return err
+		}
+
+		// The service and all external resources, if applicable, have been removed.
+		return nil
+	}
+
+	if service.DeletionTimestamp != nil {
+		// This service has been successfully deleted so there is nothing else to do.
+		return nil
+	}
+
+	if needToAddFinalizer(service) {
+		service, err = s.addFinalizer(service)
+		if err != nil {
+			return err
+		}
+
+		// Continue processing the service now that it has been updated with the finalizer.
+	}
+
+	return s.processServiceUpdate(service, key)
 }
 
 // Returns an error if processing the service deletion failed, along with a time.Duration
 // indicating whether processing should be retried; zero means no-retry; otherwise
 // we should retry after that Duration.
-func (s *ServiceController) processServiceDeletion(key string) error {
-	cachedService, ok := s.cache.get(key)
-	if !ok {
-		glog.Errorf("service %s not in cache even though the watcher thought it was. Ignoring the deletion", key)
-		return nil
+func (s *ServiceController) processServiceDeletion(service *v1.Service, key string) error {
+	err := s.processLoadBalancerDelete(service, key)
+	if err != nil {
+		return err
 	}
-	return s.processLoadBalancerDelete(cachedService, key)
+
+	return nil
 }
 
-func (s *ServiceController) processLoadBalancerDelete(cachedService *cachedService, key string) error {
-	service := cachedService.state
-	// delete load balancer info only if the service type is LoadBalancer
-	if !wantsLoadBalancer(service) {
-		return nil
-	}
+func (s *ServiceController) processLoadBalancerDelete(service *v1.Service, key string) error {
+	glog.Infof("Deleting existing load balancer for service %s that no longer needs a load balancer.", key)
 	s.eventRecorder.Event(service, v1.EventTypeNormal, "DeletingLoadBalancer", "Deleting load balancer")
 	err := s.balancer.EnsureLoadBalancerDeleted(s.clusterName, service)
 	if err != nil {
 		s.eventRecorder.Eventf(service, v1.EventTypeWarning, "DeletingLoadBalancerFailed", "Error deleting load balancer (will retry): %v", err)
 		return err
 	}
+
 	s.eventRecorder.Event(service, v1.EventTypeNormal, "DeletedLoadBalancer", "Deleted load balancer")
-	s.cache.delete(key)
 
 	return nil
 }
+
+func (s *ServiceController) addFinalizer(svc *v1.Service) (*v1.Service, error) {
+	if hasFinalizer(svc) {
+		return svc, nil
+	}
+
+	updated := svc.DeepCopy()
+	updated.ObjectMeta.Finalizers = append(svc.ObjectMeta.Finalizers, serviceLoadBalancerFinalizer)
+
+	svc, err := serviceutil.Patch(s.kubeClient.CoreV1(), svc, updated)
```

Comment:
It's because we're using  `StrategicMergePatchType`.

Just as an example, lets say the service has finalizer `foo` and we add finalizer `bar`. The patch bytes would be the following 

`"metadata":{"$setElementOrder/finalizers":["foo","bar"],"finalizers":["bar"]}}`

which only specifies the new finalizer `bar` and omits the previous finalizer `foo` since they'll be merged in the apiserver.